feature(comments): river comment links go to right page and per page configurable

mrclay opened the pull request:
Fixes #7936 and #4193. Also makes comments per page configurable by hook. Although this is already configurable by the "view_vars" hook, the hook allows the comment URLs to point to the correct page.

mrclay committed:
feature(comments): river comments/discussion replies go to right page

Fixes #7936

mrclay committed:
feature(comments): allows setting comments per page via hook

mrclay commented:
@hypeJunction curious how this would affect your Ajax work

jdalsem commented:
LGTM

mrclay commented:
FYI this adds a COUNT query for each comment in the river. A lot of our per work should help offset this.

jdalsem commented:
out of the box thought:
why not have a link to `/comment/<guid>` that gets routed to the correct entity url (plus comments offset)?

mrclay commented:
I was just typing that suggestion :)

jdalsem commented:
offset could change over time, `comment/guid` will never change (unless the comment is removed, so maybe entity link should always be passed along)

That could also fix the "read the comment" link in the mail (that is crappy now, as it only links to the entity, and not the comment)

mrclay commented:
Agreed. I guess we'd also want `/discussion/reply/GUID`.

jdalsem commented:
related to #4193

mrclay commented:
To be clear, $comment->getURL() would return `/comment/$comment_guid/$target_guid` and we should add another hook in the comment/ handler (or should we let plugin authors use the forward hook?).

mrclay committed:
chore(http): adds URLs that redirect to comments/replies in context

For comments and discussion replies, getURL() now points to URLs that
redirect to the the correct containing entity, offset, and fragment. This
eliminates the need to query this in the river and makes for safer URLs
in notifications, since the offset could potentially change.

Fixes #4193

mrclay commented on engine/lib/comments.php:
```diff
@@ -85,6 +89,67 @@ function _elgg_comments_page_handler($page) {
+			'offset' => $offset,
+		]) . "#elgg-object-{$comment->guid}";
+
+	$url = _elgg_services()->hooks->trigger('forward_url', 'object/comment', null, $url);
```
Anyone know what to call this hook?

juho-jaakkola replied:
I don't think we need this. We already have:
 - ```'entity:url', 'object'```
 - ```'forward', 'system'```

mrclay commented on engine/lib/comments.php:
```diff
@@ -85,6 +89,67 @@ function _elgg_comments_page_handler($page) {
+ * @return void
+ * @access private
+ */
+function _elgg_comment_redirect($comment_guid, $fallback_guid) {
```
These added functions are so similar but differ in so many ways I see no reason to put energy into refactoring.

mrclay commented:
I'd like to get one more core thumbs up on this idea.

juho-jaakkola commented:
Tested and LGTM

manacim commented:
+1
